arch/xtensa/src/esp32s3/esp32s3_rt_timer.c

```diff
@@ -0,0 +1,1054 @@
+   * should remove this flag.
+   */
+
+  timer->flags &= ~RT_TIMER_REPEAT;
+
+  /* Only timers in "ready" state can be stopped */
+
+  if (timer->state == RT_TIMER_READY)
+    {
+      bool ishead;
+
+      /* Check if the timer is at the head of the list */
+
+      if (timer == container_of(priv->runlist.next,
+                                struct rt_timer_s, list))
+        {
+          ishead = true;
+        }
+      else
+        {
+          ishead = false;
+        }
+
+      list_delete(&timer->list);
+      timer->state = RT_TIMER_IDLE;
+
+      if (ishead)
+        {
+          if (!list_is_empty(&priv->runlist))
+            {
+              /* Set the value from the next timer as the new hardware timer
+               * alarm value.
+               */
+
+              struct rt_timer_s *next_timer =
+                container_of(priv->runlist.next, struct rt_timer_s, list);
+
+              rt_timer_setalarm(false);
+              rt_timer_setalarmvalue(USEC_TO_CYCLES(next_timer->alarm));
+              rt_timer_setalarm(true);
+            }
+        }
+    }
+
+  spin_unlock_irqrestore(&priv->lock, flags);
+}
+
+/****************************************************************************
+ * Name: rt_timer_thread
+ *
+ * Description:
+ *   RT Timer working thread: Waits for a timeout semaphore, scans the
+ *   timeout list and processes all the timers in the list.
+ *
+ * Input Parameters:
+ *   argc          - Not used.
+ *   argv          - Not used.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success. A negated errno value is returned to
+ *   indicate the nature of any failure.
+ *
+ ****************************************************************************/
+
+static int rt_timer_thread(int argc, char *argv[])
+{
+  struct esp32s3_rt_priv_s *priv = &g_rt_priv;
+
+  while (1)
+    {
+      /* Waiting for all timers to time out */
+
+      int ret = nxsem_wait(&priv->toutsem);
+      if (ret < 0)
+        {
+          tmrerr("ERROR: Wait priv->toutsem error=%d\n", ret);
+          assert(0);
+        }
+
+      irqstate_t flags = spin_lock_irqsave(&priv->lock);
+
+      /* Process all the timers in list */
+
+      while (!list_is_empty(&priv->toutlist))
+        {
+          /* Get the first timer in the list */
+
+          struct rt_timer_s *timer = container_of(priv->toutlist.next,
+                                                  struct rt_timer_s, list);
+
+          /* Cache the raw state to decide how to deal with this timer */
+
+          enum rt_timer_state_e raw_state = timer->state;
+
+          /* Delete the timer from the list */
+
+          list_delete(&timer->list);
+
+          /* Set timer's state to idle so it can be restarted by the user. */
+
+          timer->state = RT_TIMER_IDLE;
+
+          spin_unlock_irqrestore(&priv->lock, flags);
+
+          if (raw_state == RT_TIMER_TIMEOUT)
+            {
+              timer->callback(timer->arg);
```

Comment:
Is callback mandatory or we need to make `if (timer->callback != NULL)` conditional call?

Reply:
Yes, the driver expects a valid callback function here. I've added some `DEBUGASSERT` on `esp32s3_rt_timer_create` to catch this. Thanks.